Add E2E test for guest cluster IPAM

tuommaki opened the pull request:
Guest cluster IPAM implementation is renewed architecturally as
kubernetesd is going to be removed. In order to make sure that allocated
networks are not in use, refactored IPAM implementation must be taking
cloud provider state into account and therefore new implementation
resides in provider specific operator (i.e. aws-operator, azure-operator
and kvm-operator). This test is for verifying that these implementations
cover basic use cases.

tuommaki committed:
Add E2E test for guest cluster IPAM

Guest cluster IPAM implementation is renewed architecturally as
kubernetesd is going to be removed. In order to make sure that allocated
networks are not in use, refactored IPAM implementation must be taking
cloud provider state into account and therefore new implementation
resides in provider specific operator (i.e. aws-operator, azure-operator
and kvm-operator). This test is for verifying that these implementations
cover basic use cases.

xh3b4sd commented on ipam/ipam.go:
```diff
@@ -0,0 +1,128 @@
+
+	// First create all three guest clusters
+	for _, cn := range clusters {
+		err = c.provider.RequestGuestClusterCreation(cn)
```
Why not just `CreateCluster`? I also would like to have logging before and after the action so this would be more self explanatory in consequence.

tuommaki replied:
`Request...` because I wanted to emphasize the fact that this function will not wait for creation but only request for a new cluster (which creation of provider specific config kind of is). I don't have strong opinion here though, if you think `CreateCluster` is good despite the fact it's not waiting for cluster creation.

I'll add logging.

xh3b4sd replied:
Lets rather have godoc on the provider interface and if necessary in the code here to make certain things clear. The function name cannot do this job I think.

xh3b4sd commented on ipam/ipam.go:
```diff
@@ -0,0 +1,128 @@
+
+	// TODO: Verify subnet properties for three created clusters.
+
+	// Now terminate guest cluster #2 and create guest cluster #4.
```
Same with the logging. Instead of comments we could have logging which explains what is going on.

tuommaki replied:
Ack.

xh3b4sd commented on ipam/provider/aws.go:
```diff
@@ -0,0 +1,144 @@
+// format string value. ClusterName is not based on env variable because there
+// would be too high risk for unintentional disruption of other tests if
+// ${CLUSTER_NAME} would be dynamically changed on the fly.
+const apiextensionsAWSConfigE2EChartValues = `commonDomain: ${COMMON_DOMAIN}
```
Why do we have to have this here? I thought there is some place in which we hold all of them. Also the whole environment magic is again somewhere in the depths in the code and I am strictly against that. So when we get a chance to make this better somehow I would like to do so. For instance the provider should just be configured with the necessary bits and pieces.

tuommaki replied:
I copied this here because with current implementation this is specific to my test case. Refactoring existing code would be too much work in the scope of my story as I need to progress with that. I fully understand your concern with env variables so would it be ok if I implement the same workaround here as with `e2e-harness/framework.Guest` that there is config parameters and if those are missing, env variables are looked up?

xh3b4sd replied:
Yeah this would be a step into the right direction I think.

xh3b4sd commented on ipam/provider/aws.go:
```diff
@@ -0,0 +1,144 @@
+		// NOTE we ignore errors here because we cannot get really useful error
+		// handling done. This here should anyway only be a quick fix until we use
+		// the helm client lib. Then error handling will be better.
+		framework.HelmCmd(fmt.Sprintf("delete --purge %s", deploymentName))
```
Same here as with the env vars. Lets stop doing the shell out bullshit. We can just use a helm client. We have the library. We have the frameworks. That is all there and set up. And then we also have some proper error handling. 

I also think that the resource creation should be abstracted somehow. We do that the same way like in all e2e tests so there is some common ground to cover.

tuommaki replied:
Can you elaborate, which helm client and are there already examples how to use it? This piece of code was from `aws-operator's` present e2e-test setup: https://github.com/giantswarm/aws-operator/blob/master/integration/setup/setup.go#L172-L175

Regarding abstraction here, working on e2e test framework would be too large task within the scope of my current story although I do agree with you regarding interface here.

xh3b4sd replied:
The guest and hist frameworks already expose a k8s and a g8s client. They have a rest config. The frameworks should then also just expose a helm client which is also set up with the rest. The necessary magic is already done in https://github.com/giantswarm/guestcluster.

xh3b4sd replied:
Using the helm client to delete a chart is done in `chart-operator`. See https://github.com/giantswarm/chart-operator/blob/9d0955af54e11dbd5c6661acf8ac4ecc3a92b4b1/service/controller/v3/resource/chart/delete.go.

tuommaki replied:
I totally understand you and agree here, but this change is pretty big and should be done as a separate project in SIG Operator or SIG Releng in order to cover relevant fields in `e2e-harness`. For the scope of this PR, in order to get `aws-operator` IPAM implementation eventually ready, I'd rather leave this as-is.

xh3b4sd replied:
Can we have a TODO issue for these things so we track the problem and its addressed progress with all the places where we have to fix it.

tuommaki replied:
Yes. Here: https://github.com/giantswarm/giantswarm/issues/4091

xh3b4sd commented on ipam/spec.go:
```diff
@@ -0,0 +1,21 @@
+	//  - Create guest clusters #1, #2, #3.
+	//	- Wait for guest clusters to be ready.
+	//	- Verify that clusters have distinct subnets.
+	//  - Terminate guest cluster #2 and create guest cluster #4.
```
Immediately. This is the key aspect which we want to verify because this is the case which fails in production right now.

tuommaki replied:
Ack.

xh3b4sd commented on ipam/spec.go:
```diff
@@ -0,0 +1,21 @@
+	//  - Terminate guest cluster #2 and create guest cluster #4.
+	//  - Wait for guest clusters to be deleted and created.  - Verify that
+	//	  clusters have distinct subnets and created cluster #4 did not receive
+	//	  same subnet that deleted cluster #2 had.
```
I am not sure what happened to the format here. 

<img width="217" alt="mha_hart_mah_sole" src="https://user-images.githubusercontent.com/552769/42504662-be64cd64-843c-11e8-9d18-a42db09104d4.png">

tuommaki replied:
This is probably result of my way of editing this comment here, Vim automatic creation of new comment lines + `gofmt` which ended with mixed tabs and spaces. Fixing... :)

tuommaki committed:
Fix formatting + emphasize immediacy when terminating #2 and creating #4

tuommaki committed:
Remove GuestFramework from config as it's unneeded

tuommaki committed:
Address review: Rename functions, add logging

tuommaki committed:
Use Go text/template instead of environment variables

In order to modify contents of AWSConfig chart values file, create Go
text/template for it, accept parameters as fields in config and in case
those are missing, get them from environment variables.

tuommaki committed:
Rename impl function names as well

kopiczko commented on ipam/ipam.go:
```diff
@@ -0,0 +1,134 @@
+
+	{
+		guest := guestFrameworks[clusterTwo]
+		err = guest.WaitForAPIDown()
```
Is it important if API is up or down? We should only care about allocated networks.

tuommaki replied:
No, not specifically. This was more of a shortcut for waiting VPC to come up / start going down. I might reconsider to change these to actual Subnet/VPC polling, but the problem is that working with E2E tests take already too much time. Your question is reasonable though when considering test execution time.

tuommaki replied:
To return here: During development I wished I could do this, but guest VPC info is not exposed and therefore I had to settle with API state.

kopiczko commented on ipam/ipam.go:
```diff
@@ -0,0 +1,134 @@
+
+	c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("terminating guest cluster %s and immediately creating new guest cluster %s", clusterTwo, clusterFour))
+
+	c.provider.DeleteCluster(clusterTwo)
```
I think deleting clusterTwo and clusterThree would make even bigger chance of overlapping.

tuommaki replied:
True. When I initially were thinking about this, I had a plan to later add further checks later in the chain so that's why I didn't delete all of them, but due to issues with getting E2E test working at all, I got distracted, Might change this depending on decision on this.

tuommaki committed:
Update e2e-harness dep

tuommaki committed:
Merge branch 'master' into ipam-e2e-tests

tuommaki committed:
Vendor update

tuommaki committed:
Revert "Vendor update"

This reverts commit 48192913d21fa3a7f971b1e3b1634d9bd4a3a02f.

tuommaki committed:
Update backoff to use our version

tuommaki committed:
Fix variable name from awsConfig chart template

tuommaki committed:
Fix some fields from awsconfig chart template

tuommaki committed:
Fix test cluster chart deployment name

tuommaki committed:
Add required CommonDomain field to GuestConfig

tuommaki committed:
Install cert resources for spawned guests

tuommaki committed:
Set route tables for AWSConfig

tuommaki committed:
Fetch VPC PeerID only when creating new guests

tuommaki committed:
Update error format to current style, verify test properties

tuommaki committed:
Remember to initialize the map..

tuommaki committed:
Add subnet overlapping test to ipam

tuommaki committed:
Change comments to debug logs so it can be followed during execution

tuommaki committed:
Use ipam master branch

tuommaki commented on ipam/provider/aws.go:
```diff
@@ -0,0 +1,258 @@
+	framework.HelmCmd(fmt.Sprintf("delete --purge %s", deploymentName))
+}
+
+func (aws *AWS) installCertResources(clusterName string) error {
```
This is pretty much a modified copy from `e2e-harness` where it was in a bit different form. I'm unsure if this should be left as-is here, or the `e2e-harness` modified to provide this feature? This test is probably the only one needing this kind of feature for now so hence I settled with copy for now.

tuommaki reviewed (commented):
Now this should be pretty much what I expect it to be. Please take a new look on this.

tuommaki committed:
Merge branch 'master' into ipam-e2e-tests

tuommaki committed:
dep ensure

xh3b4sd commented on ipam/provider/spec.go:
```diff
@@ -0,0 +1,18 @@
+	// cluster. Provider specific operator reconciles on it and deletes the
+	// guest cluster eventually. This function does not wait for guest cluster
+	// to get deleted but only deletes the chart deployment of provider config.
+	DeleteCluster(clusterName string)
```
IMO this method should return an error and we should always handle it somehow so we do not magically ignore information in case shit hits the fan.

tuommaki replied:
I can make this return an error, but especially in test clean-up face there's nothing that can be done if deletion fails.

xh3b4sd replied:
We can log the error.

xh3b4sd commented on ipam/ipam.go:
```diff
@@ -0,0 +1,223 @@
+
+	c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("terminating guest cluster %s and immediately creating new guest cluster %s", clusterTwo, clusterFour))
+
+	c.provider.DeleteCluster(clusterTwo)
```
The test steps are still kind of wild and it is pretty hard to read and thus understand the test and how it works. So far we have a pretty clear layout where we have the `{ }` code blocks and opening and closing logs so all functionality is described and encapsulated. I miss this here and would appreciate to continue with this consistently. It makes it way more easy to read and understand the test and makes it less "wild" and "messy".

tuommaki replied:
Roger. I'll work on this.

xh3b4sd commented on ipam/provider/aws.go:
```diff
@@ -0,0 +1,258 @@
+  routeTable0: {{ .AWSRouteTable0 }}
+  routeTable1: {{ .AWSRouteTable1 }}
+  vpcPeerId: {{ .AWSVPCPeerID }}
+`
```
Don't we have these things in the `apiextensions` repo? I would like to try defining these things in one place and not over and over again here and there.

tuommaki replied:
There's actually comment describing that this originates from `e2etemplates`, but this test requires changes so hence copying. I'll see if I could somehow incorporate these to source repo so it would be only in one place.

xh3b4sd commented on ipam/provider/aws.go:
```diff
@@ -0,0 +1,258 @@
+			fName := f.Name()
+			err = f.Close()
+			if err != nil {
+				// XXX: Tempfile leak.
```
What is up with that?

tuommaki replied:
It's kind of a warning what would inevitably happen when above thing fails. If you think it's misleading, I can remove it. I just thought that it would be good to explicitly note what would happen if execution branches to `if`-body here.

xh3b4sd replied:
I may just not used to the brief expression as it is. I usually prefer more words.

tuommaki reviewed (commented):
Thanks for feedback, I'll start working on these.

tuommaki committed:
Address review comments

tuommaki commented on Gopkg.toml:
```diff
@@ -83,3 +87,7 @@ required = [
+
+[[constraint]]
+  branch = "master"
+  name = "github.com/giantswarm/e2etemplates"
```
Note that this is now imported in here as well (as it is in the project that imports this library).

rossf7 commented on ipam/ipam.go:
```diff
@@ -0,0 +1,252 @@
+	)
+
+	defer func() {
+		c.logger.LogCtx(ctx, "level", "debug", "message", "ensuring all guest clusters possibly created in test are deleted.")
```
Nitpicky but I don't think the possibly is needed. For me its still clear without it and easier to read.

tuommaki replied:
Check. Removed.

rossf7 commented on ipam/ipam.go:
```diff
@@ -0,0 +1,252 @@
+		}
+	}()
+
+	// clusters to create in first batch
```
Sorry again nitpicking about the comments but can you make them all sentences.

tuommaki replied:
Sure.

rossf7 commented on ipam/ipam.go:
```diff
@@ -0,0 +1,252 @@
+			if err != nil {
+				return microerror.Mask(err)
+			}
+			c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("create guest cluster %s", cn))
```
created

tuommaki replied:
Yup :)

rossf7 commented on ipam/ipam.go:
```diff
@@ -0,0 +1,252 @@
+	}
+
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("waiting for guest cluster %s to become up", clusterFour))
```
come up

tuommaki replied:
:+1:

rossf7 commented on ipam/provider/aws.go:
```diff
@@ -0,0 +1,259 @@
+	o := func() error {
+		// NOTE we ignore errors here because we cannot get really useful error
+		// handling done. This here should anyway only be a quick fix until we use
+		// the helm client lib. Then error handling will be better.
```
Ack makes sense and I like that we're moving everything to helmclient.

tuommaki replied:
Oh, yeah, this comment is kind of true, but there is actually error propagation now. Removed the comment.

rossf7 commented on ipam/provider/aws.go:
```diff
@@ -0,0 +1,259 @@
+			err = f.Close()
+			if err != nil {
+				// XXX: Tempfile cannot be removed due to error. It must be
+				// removed manually.
```
For my interest. Do we need to remove this file manually when running tests locally? I like that the filename is logged.

tuommaki replied:
Yes. I made the error message more explicit about this.

rossf7 reviewed (approved):
LGTM. Just some minor things on comments and logs. 

Great that we have e2e coverage for IPAM and I like how e2etests is coming together as a library.

tuommaki committed:
Address review comments

tuommaki commented:
Need to check `aws-operator` e2e execution next. There's something wrong there, so this is not 100% complete yet, but close.

kopiczko commented on ipam/provider/error.go:
```diff
@@ -0,0 +1,39 @@
+}
+
+var tooManyResultsError = &microerror.Error{
+	Kind: "too many results",
```
Should it be `tooManyResults`?

tuommaki replied:
Oh, perhaps. Same probably above. This got changed in the middle so I kind of short sighted here. Good catch.

kopiczko commented on ipam/provider/aws.go:
```diff
@@ -0,0 +1,256 @@
+			VersionBundleVersion:      getValue(config.VersionBundleVersion, "VERSION_BUNDLE_VERSION"),
+		},
+		hostClusterName: getValue(config.ClusterName, "CLUSTER_NAME"),
+	}
```
I think it would be good to do all the env stuff in separate `env.go`.

tuommaki replied:
Roger. I'll refactor this.

kopiczko commented on ipam/provider/aws.go:
```diff
@@ -0,0 +1,256 @@
+
+func certDeploymentName(clusterName string) string {
+	return fmt.Sprintf("%s-cert-config-e2e", clusterName)
+}
```
Nit: I think it would be nice if both deployment names started or ended with the clusterName.

tuommaki replied:
Roger. I'll change.

kopiczko commented on ipam/provider/aws.go:
```diff
@@ -0,0 +1,256 @@
+			}
+
+			return nil
+		}
```
Can't be `Host.InstallCertResource()` used here?

tuommaki replied:
No because it doesn't accept cluster name. I was thinking if that should be refactored, but then again, there is a ticket in SIG Operator to refactor all helm stuff towards client library instead of using command line escape so I think it can be changed at that point.

kopiczko commented on ipam/provider/aws.go:
```diff
@@ -0,0 +1,256 @@
+		}
+
+		return nil
+	}
```
I think this can be simplified. Please check https://github.com/giantswarm/aws-operator/blob/d5994596e696081fbb693e45e6e3585004853df9/integration/setup/setup.go#L137-L170

tuommaki replied:
Good shout. I will.

tuommaki replied:
Uff.. That `err = config.Host.InstallResource("apiextensions-aws-config-e2e", values, ":stable")` uses hard coded deployment name. I would need to refactor that and fix existing e2e tests elsewhere. I would rather not do that at this point. If you don't mind, I'd rather leave this as-is as otherwise this rabbit hole never ends. WDYT?

kopiczko replied:
Should we have TODO + issue then? I think it would be good to gather features we need to incorporate into e2e-harness and them implement them. Otherwise those tests will become hard to read and maintain.

tuommaki replied:
Similar to certconfig installation, I see this as part of the task when refactoring towards use of HelmClient instead of executing commands. I just had a chat with @MarcelMue and implementation for helm client is actually in the works currently and it seems good idea to make changes there for the API in order to incorporate these requirements and then make the switch. TODO for that is here: https://github.com/giantswarm/giantswarm/issues/4091

kopiczko commented on ipam/ipam.go:
```diff
@@ -0,0 +1,250 @@
+		for _, cn := range []string{clusterOne, clusterTwo, clusterThree, clusterFour} {
+			err := c.provider.DeleteCluster(cn)
+			if err != nil {
+				c.logger.LogCtx(ctx, "level", "error", "message", fmt.Sprintf("cluster %s deletion failed: %#v", cn, err))
```
I'd be good to log `stack` here IMO.

tuommaki replied:
Ack. I'll add.

kopiczko commented on ipam/ipam.go:
```diff
@@ -0,0 +1,250 @@
+	clusters := []string{clusterOne, clusterTwo, clusterThree}
+
+	{
+
```
Nit: newline

tuommaki replied:
Check.

kopiczko commented on ipam/ipam.go:
```diff
@@ -0,0 +1,250 @@
+			}
+
+			guestFrameworks[cn] = guestFramework
+			err = guestFramework.Setup()
```
Does it make sense to run it in parallel?

tuommaki replied:
I don't think it matters here. The above loop with calls to `CreateCluster()` creates `AWSConfigs` which triggers cluster creation in parallel. This is only checking that API is up. It would need to wait for the last one anyway so I don't see major gain in speed but it would be a bit more complicated to read.

kopiczko commented on ipam/ipam.go:
```diff
@@ -0,0 +1,250 @@
+		c.logger.LogCtx(ctx, "level", "debug", "message", "verifying that subnet allocations don't overlap")
+
+		for _, cn := range clusters {
+			awsConfig, err := c.hostFramework.AWSCluster(cn)
```
This looks provider specific.

tuommaki replied:
Yes, this is. I made a note somewhere earlier that this first implementation is indeed provider specific. I'll refactor this when there's another provider taking this into use. That way I know what's actually needed and can create better abstractions.

kopiczko commented on ipam/ipam.go:
```diff
@@ -0,0 +1,250 @@
+	return err
+}
+
+func verifyNoOverlap(subnet1, subnet2 string) error {
```
Does it make sense to unit test it?

tuommaki replied:
Unsure. `ipam.Contains()` has tests and this function is pretty simple. I can do it if you see it useful here.

kopiczko reviewed (commented):
I understand this is still WIP right? I'm asking because of message in #news-product which suggested it's ready.

tuommaki reviewed (commented):
> I understand this is still WIP right?

No. I'm aiming for approval here now. I'll do fixes and make a note again then.

tuommaki committed:
Address review comments

kopiczko reviewed (approved):
LGTM

tuommaki committed:
Don't call env function unless necessary

xh3b4sd committed:
Merge branch 'master' into ipam-e2e-tests

# Conflicts:
#	Gopkg.lock
#	vendor/github.com/giantswarm/microstorage/LICENSE
#	vendor/github.com/prometheus/client_golang/prometheus/collector.go
#	vendor/github.com/prometheus/client_golang/prometheus/desc.go
#	vendor/github.com/prometheus/client_golang/prometheus/histogram.go
#	vendor/github.com/prometheus/client_golang/prometheus/http.go
#	vendor/github.com/prometheus/client_golang/prometheus/metric.go
#	vendor/github.com/prometheus/client_golang/prometheus/process_collector.go
#	vendor/github.com/prometheus/client_golang/prometheus/registry.go
#	vendor/github.com/prometheus/client_golang/prometheus/summary.go
#	vendor/github.com/prometheus/client_golang/prometheus/value.go
#	vendor/github.com/spf13/cobra/cobra/cmd/license_apache_2.go
#	vendor/k8s.io/kube-aggregator/LICENSE

xh3b4sd committed:
refactors the IPAM test

xh3b4sd committed:
updates vendor deps

xh3b4sd committed:
refactors the IPAM test

xh3b4sd commented:
I rewrote the thing a bit. Improved the provider interface. We can implement it in the provider operators themselves. That pattern might be a cool thing in general. Also to keep the dependency graph cleaner here. Nice learning. I also improved logging and the code structure to please me personally a bit more. Please have another look.

kopiczko reviewed (approved):
Nice. I like it.

xh3b4sd committed:
Merge branch 'master' into ipam-e2e-tests

xh3b4sd commented:
IMO the tests themselves are fine. Going with this now.
